### PR TITLE
more secure uuid

### LINKED ISSUE #23: Consider replacing UUID with something more secure
The code currently uses UUID identifiers to refer to uploads. UUIDs were chosen as they have two qualities:

- They're unique, preventing duplicates
- They are random (and therefore used to prevent people from guessing a value)

It seems that the last part might is not necessarily provide as much security as that was assumed when creating this code. UUID's implementation does not add _that much_ randomness into its value. See https://neilmadden.blog/2018/08/30/moving-away-from-uuids/

It should be considered to replace (or augment) the usage of UUID to improve the security aspect of things. Care should be taken to not compromise the uniqueness-aspect of the identifier that's used.

### PR DESCRIPTION
**Pull request by gmournos**
fixes #23 .
It does not take care of the uniqueness.
Looking at the java8 implementation of UUID. randomUUID, it seems that the generated uuid is type 4, so just a random number.
There is no uniqueness, at least I dont see any.

Apparently collision probability is 1 in a billion and can be ignored...
https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions

### CODE REVIEW

**Comment by guusdk** [comment]
This looks great, thanks!

I'd like to investigate the uniqueness of things for a bit - I'll get back on that.

In the mean time: how does this respond if "old" URLs are being provided? If the file storage is persisted, it would be nice to have some kind of backwards compatibility.

**Comment by gmournos** [comment]
c6a7237 should take care of legacy uuids

File: src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java [hunk 1]
```diff
@@ -25,12 +28,12 @@ public static SecureUUID generate() {
     }
     
     public static SecureUUID fromString(String id) {
-        return new SecureUUID(id);
+    	return new SecureUUID(id);
```
Comment (guusdk) [inline src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java:31]: Weird whitespace indent?

File: src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java [hunk 2]
```diff
@@ -65,5 +68,11 @@ public int hashCode() {
 		}
 		return Objects.hash(this.id);
 	}
+	
+	public static void main(String...args) throws Exception {
```
Comment (guusdk) [inline src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java:72]: Let's not have `public static void main()` in this code :) You might want to add it as a unit test instead?

File: src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java (hunk 1 above)
Comment (gmournos) [reply src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java]: ok. Changed \t to four spaces. fad8c4fd8ee402245b29122752b31e556c5ad96a

File: src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java (hunk 2 above)
Comment (gmournos) [reply src/main/java/nl/goodbytes/xmpp/xep0363/SecureUUID.java]: ok. it was trivial so I removed it. fad8c4fd8ee402245b29122752b31e556c5ad96a

**Comment by guusdk** [comment]
Thanks. Having prodded things a little, I don't believe your solution to be generating less unique values than the code it replaces.

**Pull request merged**